File: 39. Combination Sum/Solution.cpp
class Solution {
public:
    void generator(int i, vector<int> arr, int target, vector<vector<int>> &ans, vector<int> temp){
        if(i==arr.size()){
            if(target==0){
                ans.push_back(temp);

            }
            return;
        }
        if(arr[i]<=target){
            temp.push_back(arr[i]);
            generator(i,arr,target-arr[i],ans,temp);
            temp.pop_back();
        }
        generator(i+1,arr,target,ans,temp);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> te;
        generator(0,candidates,target,result,te);
        return result;
    }
};

File: 39. Combination Sum/Solution.py
class Solution:
    def generator(self,i:int, arr:List[int], target:int, ans:List[List[int]], temp:List[int])-> None:
        if i==len(arr):
            if target==0:
                ans.append(temp[:])
            return
        if arr[i]<=target:
            temp.append(arr[i])
            self.generator(i,arr,target-arr[i],ans,temp)
            temp.pop()
        self.generator(i+1,arr,target,ans,temp)

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        te = []
        self.generator(0,candidates,target,result,te)
        return result
